internal/blocktx/processor.go

```diff
@@ -188,45 +187,56 @@ func (p *Processor) StartBlockProcessing() {
 			case <-p.ctx.Done():
 				return
 			case blockMsg := <-p.blockProcessCh:
+				blockhash := blockMsg.Header.BlockHash()
+				p.logger.Info("received block", slog.String("hash", blockhash.String()))
+
 				err := p.processBlock(blockMsg)
+
 				if err != nil {
-					blockhash := blockMsg.Header.BlockHash()
-					_, errDel := p.store.DelBlockProcessing(p.ctx, &blockhash, p.hostname)
-					if errDel != nil {
-						p.logger.Error("failed to delete block processing", slog.String("hash", blockhash.String()), slog.String("err", errDel.Error()))
-					}
+					p.logger.Error("block processing failed", slog.String("hash", blockhash.String()), slog.String("err", err.Error()))
+					p.unlockBlock(p.ctx, &blockhash)
 				}
 			}
 		}
 	}()
 }
 
-func (p *Processor) StartFillGaps(peers []p2p.PeerI) {
+// unlock block for future processing
+func (p *Processor) unlockBlock(ctx context.Context, hash *chainhash.Hash) {
+	// use closures for retries
+	unlockFn := func() error {
+		_, err := p.store.DelBlockProcessing(ctx, hash, p.hostname)
+		return err
+	}
+
+	if unlockErr := withRetry(ctx, unlockFn, 5); unlockErr != nil {
+		p.logger.ErrorContext(ctx, "failed to delete block processing", slog.String("hash", hash.String()), slog.String("err", unlockErr.Error()))
+	}
+}
+
+func (p *Processor) startBlockProcessedGuard(ctx context.Context, hash *chainhash.Hash) {
```

Comment:
This is a really nice feature, but:
1. In case of an error in block processing, before that block is upserted into db, the block will be unlocked in the `processBlock` function, while this function will do 6 queries to database which will try to unlock an already unlocked block.
2. When we process a block, get an error when processing, unlock the block and then `fillGaps` will get us that block again, this goroutine may unlock that block even tho it's being processed (you found it 😄 )

For case 1. I recommend removing throwing an error when `rowsAffected != 1` in `DelBlockProcessing`. It will then stop trying to unlock an already unlocked block after the first try.

Reply:
Changed! 
Now this guardian goroutine completes when block processing finishes. Also, `unlock` no longer treats `ErrBlockNotFound` as a reason to retry